MAGNETIC_PARKING_EXTRUDER

Weruminger opened the pull request:
### Description

project: https://www.thingiverse.com/thing:3080893
Two separate X-carriages with extruders that connect to a moving part via a magnetic docking mechanism. only with movements and no solenoid.
 
The kinematic coupling and decoupling will be done properly from the Firmware.

Changes for Implementation are done in
```
Marlin/src/Marlin.cpp	
Marlin/src/inc/Conditionals_post.h	
Marlin/src/gcode/gcode.cpp	
Marlin/src/gcode/gcode.h	
```
Function realized in
```
Marlin/src/module/tool_change.cpp	
Marlin/src/module/tool_change.h	
```
Configuration Sample in 
```
Marlin/src/config/default/Configuration.h
```
Config Overwriting G-Coder (thi may be excluded or changed, because the used G-Code is not discussed)
```
Marlin/src/gcode/probe/M951.cpp	
```
~~ISSUE FW Retraction found: Did not wor correct in combination with G10/G11 and G10 S0 /  G11 S0
solved in ...~~[Cancelled to be discussed in Issue]

~~ISSUE in  Z_OFFSET handling found: it simply does not work with CURA generated gcode
solved in ...~~[Obsolete]

~~OPTIONAL Automatic Fan Switch on Boards with multiple PWM FAN Outputs ( alternative to FAN Multiplexing )~~[Moved to new PR]

### Benefits
Easy integration of Dual Parking Extrude that works with pure kinematic pickup

~~Solved Hotend offset issue ( it does not work in current version of this branch wit standard CURA gcode,~~

~~Solved FW Retraction Issue, in current version it works only if you grant a small given process that can only be refined by g-code post processing. with this change it works as described and as expected and no g-code have to be post processed to refine the needs.~~

thinkyhead commented:
Sorry for closing. This looked like a mistake because it has no title and the commit history looks like an abandoned fish market.

thinkyhead committed:
MAGNETIC_PARKING_EXTRUDER

Co-Authored-By: Weruminger <weruminger@users.noreply.github.com>

thinkyhead commented:
Rebased, squashed, and applied a myriad of fixes. There are still several issues to address. To get the updated branch you must do the following:
```
git fetch origin
git checkout bugfix-2.0.x
git reset --hard origin/bugfix-2.0.x
```

thinkyhead commented on Marlin/src/feature/fwretract.cpp:
```diff
@@ -202,6 +226,9 @@ void FWRetract::retract(const bool retracting
+    else retracted[active_extruder] = retracting;
+  #else
+    retracted[active_extruder] = retracting;              // Active extruder now retracted / recovered
   #endif
```
@Weruminger — Notice that `retracted[active_extruder] = retracting` is already done on line 224. Therefore the `else` clause and extra instance here is pointless.

Weruminger replied:
@thinkyhead  "Thank you for the very quick review, title changing and code maintainment."

I see! Line 224 is too much , it adds the default retraction|resume to the swap retraction|resume. the result does not harm anything, but is not what there is expected. In case of swap retraction|resume, only feasible with multiple extruders, there shall be a decision to just retract|resume the swap feed. 

thx. I will change this.

InsanityAutomation reviewed (commented):
Given what I saw of the design, is a `zdiff` of 0 correct? There is no guarantee that both nozzles will be identical and you may want a slight offset if working with different materials and such.

thinkyhead commented on Marlin/src/gcode/motion/G0_G1.cpp:
```diff
@@ -68,6 +70,19 @@ void GcodeSuite::G0_G1(
+          destination[Y_AXIS] -= hotend_offset[Y_AXIS][active_extruder];
+        }
+      #endif
+    #endif
```
This is a redundant and unnecessary hack. The `hotend_offset` should already have been applied to the `current_position` in `tool_change`, so that the current position already refers to the position of the selected hotend.

Weruminger replied:
Agree. 
I had an issue but it was now solved by pulling from upstream. 
I will revert this file to to head.

In future i will first discuss issues before try to solve them and i will make pull requests atomar . sorry for that.

thinkyhead commented on Marlin/src/gcode/probe/M951.cpp:
```diff
@@ -0,0 +1,41 @@
+  if (parser.seenval('L')) parkingposx[0]         = parser.value_linear_units();
+  if (parser.seenval('R')) parkingposx[1]         = parser.value_linear_units();
+  if (parser.seenval('G')) parkinggrabdistance    = parser.value_linear_units();
+  if (parser.seenval('N')) parkingslowspeed       = parser.value_linear_units();
```
The letters 'G' and 'N' are verboten as G-code parameters. You'll have to select something else.

Weruminger replied:
I'll change to G => I and N => J

Weruminger commented:
Changes made as discussed.
Rebased from downstream/bugfix-2.0.x referenced to MarlinFirmware/Marlin/bugfix-2.0.x into Fork Weruminger/Marlin/bugfix-2.0.x 
localy tested and by PR CI

InsanityAutomation commented on Marlin/Configuration.h:
```diff
@@ -193,16 +193,44 @@
+    #define MPE_SLOW_SPEED      4500      // (mm/m) Speed for last distance travel to park and couple
+    #define MPE_TRAVEL_DISTANCE   10      // (mm) Last distance point
+    #define MPE_COMPENSATION       0      // Offset Compensation -1 , 0 , 1 (multiplier) only for coupling
+    #define MPE_Z_OFFSET                  // If set, the tool-relative Z offset are subtracted on all Z axis moves (otherwise M218 will have no effect)
```
Should this be just the generic hotend z offset for the active tool?

Weruminger replied:
My fault MPE_Z_OFFSET is obsolete because G0 G1 changes have been reverted .
thx for the hint, i'll remove this line from the Cofiguration filed that contain it.

InsanityAutomation commented on Marlin/src/module/tool_change.cpp:
```diff
@@ -105,7 +105,151 @@
+    planner.buffer_line(current_position, mpe_settings.fast_feedrate, tmp_extruder);
+    planner.synchronize();
+
+    #if HAS_HOTEND_OFFSET
```
The z offset should be getting applied in the common function as it is for all other tool types. That would make this redundant.

InsanityAutomation replied:
Note the use of zdiff below to see use btw.

Weruminger replied:
Sorry i can not follow. I just implemented this feature in the same way as the previous PARKING_EXTRUDER have been implemented. can you please give me a deeper view of what you mean. then i will make the change.

Weruminger reviewed (commented):
@InsanityAutomation Thank You for the hint, i found out what you mean and i agree, this have to be changed. I'll do so

btw. the diffx diffy diffz have been provided after i started the feature coding.

InsanityAutomation commented:
Basically just remove the couple lines applying the hotend offset as it's done below. It was there before, just buried and not easy to follow, hence why I worked on reorganizing it!

Weruminger commented:
I did the changes. 
@InsanityAutomation : To the design and comentation of the code ... 👍  great stuff. i do think i had not the closest view on it. 
The given support from Scott Lahteine and You was very helpfull and incredible quick. You have my great respect for it.

InsanityAutomation commented on Marlin/src/module/tool_change.h:
```diff
@@ -53,17 +53,32 @@
-#endif // PARKING_EXTRUDER
+#elif ENABLED(MAGNETIC_PARKING_EXTRUDER)
+
+  typedef struct MPESettings {
```
@thinkyhead Do you think this is better standalone or as part of the toolchange struct and configured with M217?

Weruminger replied:
I'm not sure. too much parameter in one command and the definition of park position is different. It may confuse the user.

thinkyhead replied:
I agree that overloading `M217` with too many parameters is a bad idea.

InsanityAutomation commented on Marlin/src/config/default/Configuration.h:
```diff
@@ -193,16 +193,42 @@
+    #define MPE_SLOW_SPEED      4500      // (mm/m) Speed for last distance travel to park and couple
+    #define MPE_TRAVEL_DISTANCE   10      // (mm) Last distance point
+    #define MPE_COMPENSATION       0      // Offset Compensation -1 , 0 , 1 (multiplier) only for coupling
+    #define AUTO_FILAMENT_FAN_SELECTION   // Auto fan selection in M106/M107 for multi-extruder multi-fan solution
```
AUTO_FILAMENT_FAN_SELECTION  is not directly related to this tool type, should this be independent as well around the fan controls in config_adv?

Weruminger replied:
agree, but i do thint we need a hint here for the possibilities of fan control Fan mux and auto selection, because the need for it are multi extruder realizations. i'll put it into my ToDo List.

Weruminger replied:
done

InsanityAutomation commented:
Scott's the big one, and the decision maker on merging stuff! I just came across it because I've been spending so much time working on the tool change stuff lately with IDEX regressions! The machine type looks to get a lot of the advantages of idex with a single x gantry, just a few things lost. No duplicate mode etc. I'll dig in a bit deeper when I'm not poking between stuff at my day job, but when Scott chimes in on my last 2 notes then we can look at updating all of the default config files with these options. That'll be the next big step before it can get merged once he's satisfied with the function.

Weruminger commented:
Tested on Hardware and fixed upcomming issues.
Localy tested the kinematic on real System. 
given Parameter values measured on System.
reproduceable behaviour, 
   +fresh flashed, 
   +after changing Parameter and return to original Parameter values
   +after reset.
   +after hard reboot (switch off power)
tests i.o.
Print with dual material worked as expected. T0 /  T1

Auto Fan selection works fine. M106/M107

G10, G11, G10 S0, G11 S0 works perfect and failior safe ( retract and swap retrack only one and correct retaction state is is recognized.

Setting Parameter M851, M951, M218, M217 works as expected.

thinkyhead commented on Marlin/src/module/tool_change.cpp:
```diff
@@ -274,31 +397,16 @@
-     * 5. Grab and lock the new toolhead
-     * 6. Apply the z-offset of the new toolhead
+	 * Z Lift and Nozzle Offset shift ar defined in caller method to work equal with any Multi Hotend realization
+	 *
```
Please configure your editor to use 2 spaces for indent. Tabs are verboten!

Weruminger replied:
Sorry used VS now back to NPP that is configured in the correct way.

thinkyhead commented on Marlin/src/module/tool_change.cpp:
```diff
@@ -251,10 +378,6 @@
-    #if HAS_HOTEND_OFFSET
-      current_position[Z_AXIS] += hotend_offset[Z_AXIS][active_extruder] - hotend_offset[Z_AXIS][tmp_extruder];
-    #endif
-
```
Why is the normal solenoid-based `PARKING_EXTRUDER` code being altered in this PR?

Weruminger replied:
Same Issue as in Magnetic Parking Extruder. Z-Axis setup twice. also from caller method, because MPE was based from Parking Extruder. If wanted i'll revert this change.

thinkyhead commented on Marlin/src/module/tool_change.cpp:
```diff
@@ -274,31 +397,16 @@
-    planner.synchronize();
-
-    // STEP 2
-
```
Why is the normal solenoid-based `PARKING_EXTRUDER` code being altered in this PR?

Weruminger replied:
Same Issue as in Magnetic Parking Extruder. Z-Axis setup twice. also from caller method, because MPE was based from Parking Extruder. 
I'll revert this change and open a separate issue with pullrequest if it was accepted from the issue discussion.

thinkyhead commented on Marlin/src/module/tool_change.cpp:
```diff
@@ -398,19 +506,6 @@
-
-    #if ENABLED(DEBUG_LEVELING_FEATURE)
-      if (DEBUGGING(LEVELING)) SERIAL_ECHOLNPGM("Toolhead change done.");
-    #endif
```
Why is the normal solenoid-based `PARKING_EXTRUDER` code being altered in this PR?

Weruminger replied:
Same Issue as in Magnetic Parking Extruder. Z-Axis setup twice. also from caller method, because MPE was based from Parking Extruder. If wanted i'll revert this change.

Weruminger replied:
I did revert Changes not related to MPE from tool_change.cpp

thinkyhead replied:
If you've found an actual bug, then of course do include a fix. I just want to be sure that the changes are intentional.

thinkyhead commented on Marlin/src/feature/fwretract.cpp:
```diff
@@ -110,8 +112,32 @@ void FWRetract::retract(const bool retracting
+    if (retracted[active_extruder] == retracting) return;
+    const bool swapping = false;
+
+  #endif
```
Are you absolutely certain these changes for the `FWRETRACT` feature are necessary? There have been a lot of people working on this feature with much trial-and-error, and it would be a shame to break it after we finally settled on something that seems to be working.

Weruminger replied:
Okay I do not wan to blame anyone. I do see, there was a huge among of work on it and i do respect this. I will revert it and hold the change in my fork, because id did not work correct with my G-Code before post process the G-Code manual. Also the swap retraction had always the default retraction in addition. And with that change it works always fine with several single-nozzle multi-feeder, and with an Chimera Clone.

Is it okay if I just revert this change and open an issue to be discussed so i will create a new pull request if that is the result of the issue discussion?  

than i will follow the same procedure for all stuff that is not MPE Relevant.

I do think i shall had to follow this procedure from the beginning.

A short feedback and i will do it in this way. [already done :-)]

Weruminger replied:
I did revert fwretract.cpp

thinkyhead replied:
A new PR around the `FWRETRACT` issue would be most welcome, so that we can get more testers on it.

Weruminger replied:
I will write a detailed issue description and create a new pullrequat on it. I am not sure to provide it this Weekend, but I‘ll try.

Weruminger commented:
Is there any more action needed from my side to merge and close, Magnetic Parking Extruder Feture ? I‘m currently working on the Documentation. how this will be deployed, and who can review it. I am not a native English speaker, as you already recognized, reading my posts ;-).

InsanityAutomation commented:
Scott is currently engaged in a larger issue, and bugfix items take priority over feature updates. He'll get back to this when the endstop issue is resolved most likely.

thinkyhead commented on Marlin/src/config/default/Configuration_adv.h:
```diff
@@ -257,6 +257,21 @@
+ *      blower will cool the model. 
+ */
+
+// #define AUTO_FILAMENT_FAN_SELECTION 
```
I'm not sure we should add this option. It breaks the expectation of the default behavior of `M106`/`M107`, and we're generally concerned about having some machines out there which do one thing, while other machines do another thing given the same command. It would be better to add a flag parameter like `M106 A` that would use the current extruder index, rather than a feature that alters the default behavior of `M106`/`M107` commands.

Weruminger replied:
Okay that is feasible and it make absolute sense to introduce a flag property instead of setting up the firmware that can not be managed on runtime. I will change it. Argument name A seems to be okay for me.

Weruminger replied:
moved into an other PR #12389

Weruminger commented:
I reverted M106/107 behavour change from this PR to create a new one for it, changed as dicussed before.
https://github.com/MarlinFirmware/Marlin/pull/12389

Weruminger commented:
Modified sample Configurations to setup Magnetic Parking Extruder but disabled.

Weruminger commented:
I found out, that it is a good idea, not only lift the nozzle at tool change, but also move to a safe y position to preven oozing on the build during heat up from standby to print temperature. I would like to implement this in this PR if it is okay for you.

InsanityAutomation commented:
Look at the existing tool change park position stuff. Can probably modify it for parking types to use only y pretty easily.

Weruminger commented:
I had the idea to use the parking Parameter to set the nozzle before the change, to the safe parking y position at at the end to the safe parking x position. If the user set this position on the prime tower position, the move also save some travel time. So the movement will be done by MPE Kinematics but the Parameter where taken from the tool change parking. The concept is the same and it shall not collide, because the target position will be the same. Let me implement and test it tomorrow. Than I will see if it works as expected. I do think it is correct tu use The same Parameter for the logical same processing, also if they are realized in different modules to grant the correct order of movements and timing related to the covering process.

Weruminger commented:
On implementation i found out that, in my eyes, the order is wrong. to prevent the model it may be better to lift befor move to parking position. 

https://github.com/MarlinFirmware/Marlin/blob/81cbe9b07edfa51a8390bbf2f66ebaade254a100/Marlin/src/module/tool_change.cpp#L726

Weruminger commented:
I would decide to first close this pull request and create a new one for the safe parking position implementation. Reason: it conflicts with movements in calling method. In this case i'd leave it as it is and have a deeper view on the implementation and an MPE independen review to prevent side effects.

Weruminger commented:
Realized the safe position using G27 in extruder end script of the slicer. That works fine. important for the model safety is to stay in the safe zone till material prime is started. the Nozzle Parking feature called by G27 refines this requirement. Therfor there are no needs of code changes in MPE.

thinkyhead committed:
Add MAGNETIC_PARKING_EXTRUDER to examples

thinkyhead commented:
Rebased, squashed, replaced TABS with SPACES, and updated newly-added config examples. To get the updated branch:
```
git fetch origin
git checkout bugfix-2.0.x
git reset --hard origin/bugfix-2.0.x
```

Weruminger commented:
Okay, I did reset and merged all MPE Changes.
@thinkyhead Thanks for the changes in naming to clear separate Parking Extruder and Magnetic Parking Extruder. And thank you a lot for the great support.


i'd a look for tab's and currently found only tabs/space mixed in many h files under F:\Marlin\Marlin\src\HAL\HAL_DUE\usb\
to change this is only one click. shall i do this or shall this be done with an additional PR ?

what are the next steps?

I started Documentation for the feature in MarlinDocumentation. It may need some days, I've some workload in my company. But a.s.a.p. I'll create a pull request for that.

Weruminger commented:
Any expectations?
